File: src/two_body_simulation.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef struct {
    double x;
    double y;
} Vector;

typedef enum {
    RUNGE_KUTTA,
    EULER,
} OdeMethod;

typedef struct {
    double mass;
    Vector position;
} Planetary;

typedef struct {
    Planetary a;
    Planetary b;
    Vector relative_velocity;
} TwoBodyModel;

typedef struct {
    double eccentricity;
    double mass_ratio;
    unsigned int period;
    double delta_time;
    OdeMethod method;
} TwoBodyController;

typedef void (*OdeFunction)(TwoBodyController *controller, TwoBodyModel *model);

double dot(Vector *lhs, Vector *rhs);
double *get_vector_component(Vector *v, size_t index);
Vector add(Vector *lhs, Vector *rhs);
Vector sub(Vector *lhs, Vector *rhs);
Vector mul_scalar(Vector *v, double s);
double length(Vector *v);
Vector unit(Vector *v);
Vector direction(Vector *lhs, Vector *rhs);

// x'' = -x ( 1 + q ) / r^3
// y'' = -y ( 1 + q ) / r^3
void runge_kutta(
    TwoBodyController *controller,
    TwoBodyModel *model
);
void euler(TwoBodyController *controller, TwoBodyModel *model);

int main(void) {
    TwoBodyController controller;

    printf("Enter period (T): ");
    scanf("%u", &controller.period);
    printf("Enter delta time (dt): ");
    scanf("%lf", &controller.delta_time);
    printf("Enter mass ratio: ");
    scanf("%lf", &controller.mass_ratio);
    printf("Enter eccentricity: ");
    scanf("%lf", &controller.eccentricity);

    TwoBodyModel model;
    model.a.mass = 1.0;
    model.b.mass = controller.mass_ratio;

    model.a.position.x = model.b.mass / (model.a.mass + model.b.mass);
    model.a.position.y = 0;

    model.b.position.x = -model.a.mass / (model.a.mass + model.b.mass);
    model.b.position.y = 0;

    model.relative_velocity.x = 0;
    model.relative_velocity.y = sqrt((1 + controller.mass_ratio) * (1 + controller.eccentricity));

    OdeFunction ode = 0;
    while (!ode) {
        printf("0. Runge-Kutta\n1. Euler's Method\nEnter method: ");
        scanf("%u", &controller.method);
        if (controller.method == RUNGE_KUTTA) {
            ode = runge_kutta;
        } else if (controller.method == EULER) {
            ode = euler;
        } else {
            printf("Wrong input.\n");
        }
    }

    FILE *f = fopen("vertex.buf", "w");
    if (f) {
        for (double t = 0; t < controller.period; t += controller.delta_time) {
            fprintf(f,
                "(%.5lf, %.5lf)         ",
                model.a.position.x,
                model.a.position.y
            );
            fprintf(f,
                "(%.5lf, %.5lf)\n",
                model.b.position.x,
                model.b.position.y
            );

            ode(&controller, &model); 
        }
        fclose(f);
    }
    return 0;
}

// Passed the test
double dot(Vector *lhs, Vector *rhs) {
    double x = lhs->x * rhs->x;
    double y = lhs->y * rhs->y;

    return x + y;
}

double *get_vector_component(Vector *v, size_t index) {
    if (index == 0) {
        return &v->x;
    } else if (index == 1) {
        return &v->y;
    } else {
        exit(23);
    }
}

Vector add(Vector *lhs, Vector *rhs) {
    double x = lhs->x + rhs->x;
    double y = lhs->y + rhs->y;

    Vector result = { x, y };
    return result;
}

void add_assign(Vector *lhs, Vector *rhs) {
    lhs->x += rhs->x;
    lhs->y += rhs->y;
}

// Passed the test
Vector sub(Vector *lhs, Vector *rhs) {
    double x = lhs->x - rhs->x;
    double y = lhs->y - rhs->y;

    Vector result = {
        x,
        y
    };

    return result;
}

Vector mul_scalar(Vector *v, double s) {
    double x = v->x * s;
    double y = v->y * s;

    Vector result = { x, y };
    return result;
}

// Passed the test
double length(Vector *v) {
    return sqrt(pow(v->x, 2) + pow(v->y, 2));
}

// Passed the test
Vector unit(Vector *v) {
    double len = length(v);

    double x = v->x / len;
    double y = v->y / len;

    Vector result = { x, y };

    return result;
}

Vector direction(Vector *lhs, Vector *rhs) {
    Vector a = sub(lhs, rhs);
    Vector result = unit(&a);

    return result;
}

void runge_kutta(
    TwoBodyController *controller,
    TwoBodyModel *model
) {
    // Slope of r vector
    Vector k_r[4];
    // Slope of v vector
    Vector k_v[4];

    // Position Vector: r_b - r_a
    Vector r;
    // Velocity Vector: v_b - v_a
    Vector v;

    for (size_t i = 0; i < 4; i++) {
        if (i == 0) {
            r = sub(&model->b.position, &model->a.position);
            v = model->relative_velocity;
        } else {
            double coef;
            if (i == 1 || i == 2) {
                coef = 0.5;
            } else {
                coef = 1.0;
            }

            Vector r0 = sub(&model->b.position, &model->a.position);

            v.x = k_v[i - 1].x * (controller->delta_time * coef) + model->relative_velocity.x;
            v.y = k_v[i - 1].y * (controller->delta_time * coef) + model->relative_velocity.y;

            r.x = v.x * (controller->delta_time * coef) + r0.x;
            r.y = v.y * (controller->delta_time * coef) + r0.y;
        }
        double r_magnitude = length(&r);

        for (size_t j = 0; j < 2; j++) {
            *get_vector_component(k_v + i, j) =
                -*get_vector_component(&r, j)
                * (1 + controller->mass_ratio)
                / pow(r_magnitude, 3.0);

            *get_vector_component(k_r + i, j) = *get_vector_component(&v, j);
        }
    }

    Vector r0 = sub(&model->b.position, &model->a.position);
    for (size_t i = 0; i < 2; i++) {
        *get_vector_component(&r0, i) +=
            (1.0 / 6.0 * controller->delta_time)
                * (*get_vector_component(k_r, i)
                + 2.0 * *get_vector_component(k_r + 1, i)
                + 2.0 * *get_vector_component(k_r + 2, i)
                + *get_vector_component(k_r + 3, i));

        *get_vector_component(&model->a.position, i) =
            -(model->b.mass / (model->a.mass + model->b.mass))
            * *get_vector_component(&r0, i);

        *get_vector_component(&model->b.position, i) =
            (model->a.mass / (model->a.mass + model->b.mass))
            * *get_vector_component(&r0, i);

        *get_vector_component(&model->relative_velocity, i) +=
            (1.0 / 6.0 * controller->delta_time)
                * (*get_vector_component(k_v, i)
                + 2.0 * *get_vector_component(k_v + 1, i)
                + 2.0 * *get_vector_component(k_v + 2, i)
                + *get_vector_component(k_v + 3, i));
    }
}

void euler(
    TwoBodyController *controller,
    TwoBodyModel *model
) {
    Vector r = sub(&model->b.position, &model->a.position);
    double r_magnitude = length(&r);

    Vector acceleration = mul_scalar(&r, -(1 + controller->mass_ratio) / pow(r_magnitude, 3.0));
    Vector speed = model->relative_velocity;

    Vector new_position = mul_scalar(&speed, controller->delta_time);
    // r += v * t
    add_assign(&r, &new_position); 

    Vector new_speed = mul_scalar(&acceleration, controller->delta_time);
    // v += a * t
    add_assign(&model->relative_velocity, &new_speed);

    model->a.position = mul_scalar(&r, -(model->b.mass / (model->a.mass + model->b.mass)));
    model->b.position = mul_scalar(&r, (model->a.mass / (model->a.mass + model->b.mass)));
}

File: src/two_body_simulation.py
from dataclasses import dataclass
import math

@dataclass(repr = False)
class Vector:
    x: float
    y: float

    def __repr__(self):
        return "({:.5f}, {:.5f})".format(self.x, self.y)


    def __add__(self, other):
        return Vector(self.x + other.x, self.y + other.y)


    def __sub__(self, other):
        return Vector(self.x - other.x, self.y - other.y)


    def __mul__(self, value: float):
        return Vector(self.x * value, self.y * value)


    def __rmul__(self, value: float):
        return Vector(self.x * value, self.y * value)


    def __mul__(self, value: int):
        return Vector(self.x * value, self.y * value)


    def __rmul__(self, value: int):
        return Vector(self.x * value, self.y * value)


    def __truediv__(self, value: float):
        return Vector(self.x / value, self.y / value)


    def __iadd__(self, other):
        self.x += other.x
        self.y += other.y
        return self


    def __neg__(self):
        return Vector(-self.x, -self.y)


    def length(self):
        return math.sqrt(self.x * self.x + self.y * self.y)

class Planetary:
    def __init__(self, mass: float, position: Vector):
        self.mass = mass
        self.position = position


class TwoBodyModel:
    def __init__(self, a: Planetary, b: Planetary, relative_velocity: Vector):
        self.a = a
        self.b = b
        self.relative_velocity = relative_velocity


class TwoBodyController:
    def __init__(self, period: int, delta_time: float, eccentricity: float, mass_ratio: float, method: int):
        self.period = period
        self.delta_time = delta_time
        self.eccentricity = eccentricity
        self.mass_ratio = mass_ratio
        self.method = method

    def __init__(self):
        self.period = int(input("Enter period (T): "))
        self.delta_time = float(input("Enter delta time (dt): "))
        self.mass_ratio = float(input("Enter mass_ratio (q): "))
        self.eccentricity = float(input("Enter eccentricity: "))
        while self.eccentricity < 0 or self.eccentricity > 1:
            self.eccentricity = float(input("\nWrong input. Eccentricity must be between 0 and 1\nEnter eccentricity: "))
        self.method = int(input("\n0. Runge-Kutta\n1. Euler's Method\nEnter method: "))
        while self.method != 0 and self.method != 1:
            self.method = int(input("\nWrong input. Method must be either 0 or 1.\n0. Runge-Kutta\n1. Euler's Method\nEnter method: "))



    def init_model(self):
        m_a = 1
        m_b = self.mass_ratio

        a = Planetary(m_a, Vector(m_b / (m_a + m_b), 0))
        b = Planetary(m_b, Vector(-m_a / (m_a + m_b), 0))

        relative_velocity = Vector(0, math.sqrt((1 + self.mass_ratio) * (1 + self.eccentricity)))

        return TwoBodyModel(a, b, relative_velocity)

    
    def euler(self, model: TwoBodyModel):
        r = model.b.position - model.a.position

        a = -r * (1 + self.mass_ratio) / math.pow(r.length(), 3)
        v = model.relative_velocity

        r += v * self.delta_time
        v += a * self.delta_time

        model.a.position = -(model.b.mass / (model.a.mass + model.b.mass)) * r
        model.b.position = (model.a.mass / (model.a.mass + model.b.mass)) * r


    def runge_kutta(self, model: TwoBodyModel):
        k_r = [Vector(0, 0), Vector(0, 0), Vector(0, 0), Vector(0, 0)]
        k_v = [Vector(0, 0), Vector(0, 0), Vector(0, 0), Vector(0, 0)]

        r = Vector(0, 0)
        v = Vector(0, 0)

        for i in range(4):
            if i == 0:
                r = model.b.position - model.a.position
                v = model.relative_velocity
            else:
                coef = 0.0
                if i == 1 or i == 2:
                    coef = 0.5
                else:
                    coef = 1.0

                r0 = model.b.position - model.a.position

                v = k_v[i - 1] * (self.delta_time * coef) + model.relative_velocity
                r = v * (self.delta_time * coef) + r0


            k_v[i] = -r * (1 + self.mass_ratio) / math.pow(r.length(), 3)
            k_r[i] = v

        r0 = model.b.position - model.a.position
        r0 +=  (1 / 6 * self.delta_time) * (k_r[0] + 2 * k_r[1] + 2 * k_r[2] + k_r[3])

        model.a.position = r0 * -(model.b.mass / (model.a.mass + model.b.mass))
        model.b.position = r0 * (model.a.mass / (model.a.mass + model.b.mass))

        model.relative_velocity +=  (1 / 6 * self.delta_time) * (k_v[0] + 2 * k_v[1] + 2 * k_v[2] + k_v[3])


def main():
    controller = TwoBodyController()
    model = controller.init_model()

    count = int(controller.period / controller.delta_time) + 1
    with open('out_py.txt', 'w') as f: 
        for _ in range(count):
            f.write(str(model.a.position))
            f.write(str(model.b.position))
            f.write('\n\n')
            if controller.method == 0:
                controller.runge_kutta(model)
            else:
                controller.euler(model)


if __name__ == "__main__":
    main()
